sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -460,19 +468,70 @@
       return toBuilder().setMaxReadTime(maxReadTime).build();
     }
 
+    /**
+     * Sets {@link TimestampPolicy} to {@link TimestampPolicyFactory.LogAppendTimePolicy}.
+     * The policy assigns Kafka's log append time (server side ingestion time)
+     * to each record. The watermark for each Kafka partition is the timestamp of the last record
+     * read. If a partition is idle, the watermark advances to couple of seconds behind wall time.
+     * Every record consumed from Kafka is expected to have its timestamp type set to
+     * 'LOG_APPEND_TIME'.
+     *
+     * <p>In Kafka, log append time needs to be enabled for each topic, and all the subsequent
+     * records wil have their timestamp set to log append time. If a record does not have its
+     * timestamp type set to 'LOG_APPEND_TIME' for any reason, it's timestamp is set to previous
+     * record timestamp or latest watermark, whichever is larger.
+     *
+     * <p>The watermark for the entire source is the oldest of each partition's watermark.
+     * If one of the readers falls behind possibly due to uneven distribution of records among
+     * Kafka partitions, it ends up holding the watermark for the entire source.
+     */
+    public Read<K, V> withLogAppendTime(){
+      return withTimestampPolicyFactory(TimestampPolicyFactory.withLogAppendTime());
+    }
+
+
+    /**
+     * Sets {@link TimestampPolicy} to {@link TimestampPolicyFactory.ProcessingTimePolicy}.
+     * This is the default timestamp policy. It assigns processing time to each record.
+     * Specifically, this is the timestamp when the record becomes 'current' in the reader.
+     * The watermark aways advances to current time. If servicer side time (log append time) is
+     * enabled in Kafka, {@link #withLogAppendTime()} is recommended over this.
+     */
+    public Read<K, V> withProcessingTime() {
+      return withTimestampPolicyFactory(TimestampPolicyFactory.withProcessingTime());
+    }
+
+    /**
+     * Provide custom {@link TimestampPolicyFactory} to set event times and watermark for each
+     * partition. {@link TimestampPolicyFactory#createTimestampPolicy(TopicPartition, Optional)}
+     * is invoked for each partition when the reader starts.
+     * @see #withLogAppendTime() and {@link #withProcessingTime()}
+     */
+    public Read<K, V> withTimestampPolicyFactory(
+      TimestampPolicyFactory<K, V> timestampPolicyFactory) {
+      return toBuilder().setTimestampPolicyFactory(timestampPolicyFactory).build();
+    }
+
     /**
      * A function to assign a timestamp to a record. Default is processing timestamp.
+     * @deprecated as of version 2.4.
```

Comment:
from which version will it be removed?

Reply:
Not sure. May be in a couple of months. It is easy to switch to new API.